File: November2020/HouseRobberIII.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class HouseRobberIII {

    Map<TreeNode, Integer> cache = new HashMap<>(); // O(n)

    public int rob(TreeNode root) { // TC O(n), each node is visited once
        if(cache.containsKey(root)){
            return cache.get(root);
        }
        if(root == null){
            return 0;
        } else{
            int leftleftGrandChild  =0 ;
            int rightleftGrandChild  =0 ;

            int leftrightGrandChild  =0 ;
            int rightrightGrandChild  =0 ;

            if(root.left !=null){
                leftleftGrandChild = rob(root.left.left);
                leftrightGrandChild = rob(root.left.right);
            }
            if(root.right !=null){
                rightleftGrandChild = rob(root.right.left);
                rightrightGrandChild = rob(root.right.right);
            }

            int includingRoot = root.val + leftleftGrandChild + rightleftGrandChild + leftrightGrandChild + rightrightGrandChild;

            int leftChild = rob(root.left);
            int rightChild = rob(root.right);

            int excludingRoot = leftChild + rightChild;

            cache.put(root,  Math.max(includingRoot, excludingRoot));
            return cache.get(root);


        }
    }
}

File: December2021/C++/House_Robber_III.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int rob(TreeNode* root) {
       vector<int> res=dfs(root);
        return max(res[0],res[1]);
    }
    
    vector<int> dfs(TreeNode* root)
    {
        if(!root) return {0,0};
        vector<int> left=dfs(root->left);
        vector<int> right=dfs(root->right);
        int incurr=root->val+left[1]+right[1];
        int notcurr=max({left[0]+right[0],left[1]+right[1],left[0]+right[1],left[1]+right[0]});
        return {incurr,notcurr};
    }
};